aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp

```diff
@@ -2667,44 +2667,33 @@ void _fake_quant_per_channel_cachemask_cpu_helper(
       // write mask
       cpu_kernel(iter_mask, [=](SelfType self, float scale, scalar_t zero_point) -> bool {
         float inv_scale = 1.0f / scale;
-        const auto qval = std::lrintf(zero_point + (self * inv_scale));
+        const auto qval = std::nearbyint(zero_point + (self * inv_scale));
         return ((quant_min <= qval) && (qval <= quant_max));
       });
 
       // write fake_quant
       cpu_kernel(iter, [=](SelfType self, float scale, scalar_t zero_point) -> SelfType {
         float inv_scale = 1.0f / scale;
-        // NOLINTNEXTLINE(bugprone-narrowing-conversions,cppcoreguidelines-narrowing-conversions)
-        return (std::fmin(
-                    std::fmax(
-                        std::lrintf(zero_point + self * inv_scale),
-                        quant_min),
-                    quant_max) -
-                zero_point) *
-            scale;
+        const auto qval = std::nearbyint(self * inv_scale + zero_point);
+        const auto bounded_qval = std::fmin(quant_max, std::fmax(quant_min, qval));
+        return (bounded_qval - zero_point) * scale;
       });
     });
 
   } else {
       // write mask
       cpu_kernel(iter_mask, [=](SelfType self, float scale, int32_t zero_point) -> bool {
         float inv_scale = 1.0f / scale;
-        const auto qval = static_cast<int64_t>(zero_point + std::nearbyint(self * inv_scale));
+        const auto qval = std::nearbyint(zero_point + (self * inv_scale));
         return ((quant_min <= qval) && (qval <= quant_max));
       });
 
       // write fake_quant
       cpu_kernel(iter, [=](SelfType self, float scale, int32_t zero_point) -> SelfType {
         float inv_scale = 1.0f / scale;
-        // NOLINTNEXTLINE(bugprone-narrowing-conversions,cppcoreguidelines-narrowing-conversions)
-        return (std::fmin(
-                    std::fmax(
-                        static_cast<int64_t>(
-                            zero_point + std::nearbyint(self * inv_scale)),
-                        quant_min),
-                    quant_max) -
-                zero_point) *
-            scale;
+        const auto qval = std::nearbyint(self * inv_scale + zero_point);
+        const auto bounded_qval = std::fmin(quant_max, std::fmax(quant_min, qval));
```

Comment:
I think you may want to cast `qval` to int32/64 (I think thats what you said in a comment). quant_max/min shall also be int. So
```
const auto qval = static_cast<int32_t>(std::nearbyint(self * inv_scale + zero_point));
int32_t quant_max/min (with appropriate value);
int32_t bounded_qval = std::min(quant_max, std::max(quant_min, qval));
return static_cast<float>((static_cast<int64_t>(bounded_qval) - zero_point)) * scale;
```

Note, I used int32_t to avoid over/underflow in `bounded_qval - zero_point`.

BTW motivation to do this is to accurately approximate quant/dequant process where quant goes into int domain and in dequant you have to be in float domain during multiplication with scale.

Sorry, for keep adding more comments, but since this is something used everywhere, better to a) get it right and b) not diverge a lot from older behavior except to fix overflow/underflow issues. Although a is probably more important.

Reply:
I don't mind the comments, proves it is a proper review and the discussion is important to ensure it is actually what is desired.

I think the change as-is is correct: It is basically the same code, I only have factored out `qval` to make it easier to read and removed the cast to `int64_t` (i.e. using `const auto qval = static_cast<int64_t>(std::nearbyint(self * inv_scale + zero_point));` would be **exactly** the same code as before) See the changes to the CUDA kernels below where this is more obvious

The point here is to do the cast **after** the clipping or we can over/underflow. From the test code we have this example: `self=9.223372e+14`, `scale=1e-4`, `zero_point=0`, 4bit quantization (`qmin=0`, `qmax=15`)

So casting `self * inv_scale + zero_point` to `int32_t` will overflow even worse than the original code which at least used `int64_t`. So we need the `fmin/fmax`-clipping first which ensures we can safely cast `bounded_qval` to `int64_t` (the type of `quant_min/max`)

Yes technically `bounded_qval - zero_point` can also over/underflow, but that means `zero_point` is badly chosen, which could be fixed in user code, contrary to the issue fixed here.

> quant_max/min shall also be int. 

It is already, actually `int64_t`, which I don't really understand why 64 bit as I'd assume it would be `0 and 2^N` or `-2^(N-1) and 2^(N-1)` with N being small (4 or 8, maybe 16). Isn't it?

> BTW motivation to do this is to accurately approximate quant/dequant process where quant goes into int domain and in dequant you have to be in float domain during multiplication with scale.

We could do:
```
const auto bounded_qval = static_cast<int64_t>(std::fmin(quant_max, std::fmax(quant_min, qval)));
return (bounded_qval - zero_point) * scale;
```
- This always works assuming `quant_min/max` to losslessly convert to `double` (which is reasonable) and hence the integer `bounded_qval` must be in range `[quant_min, quant_max]` (BTW: it might make sense to convert `quant_min/max` to the type of `qval` to avoid a needless float->double promotion due to possibly calling `fmin(float, in64_t)`)
- This way `bounded_qval - zero_point` happens in integer domain and is hence exact although I don't expect this to be an issue
- However this introduces a conversion from float to int and back which might hurt or disable vectorization

So I don't see any advantage in this especially as I can't imagine a practical case where `(static_cast<int64_t>(bounded_qval) - zero_point) * scale` would produce a different value than `(bounded_qval - zero_point) * scale`:
- `bounded_qval` is exactly representable as an `int64_t` so the conversion doesn't change its value
- the latter converts `zero_point` to a `float` or `double` which may change it's value
- but even if that would be the case I'd assume the same happens in the first case before multiplying with `scale`, i.e. `float(int64_t(bounded_qval) - zero_point) == `(bounded_qval - float(zero_point))`

I'm not 100% sure though if there could be relevant rounding errors for `float(zero_point)` that do (or do not) happen for `float(bounded_qval - zero_point)` It might be possible that one can construct such a case. However note that the change of casting `bounded_qval` to int would be different to what it currently does where `zero_point` is implicitly promoted to a float type.

Reply:
I am digesting what you said (sorry was on PTO so taking longer than expected to respond), but at high level it makes sense.

My understanding is that if we somehow detect overflow/underflow and apply clipping, it would be ok?

My one concern is what you said about rounding that happens on zero point, although the code snippet I have does not quite address that.

Also your argument on bounded_qval representation in int64_t makes sense assuming qmin/qmax are exactly representable in int64_t as well.

Reply:
> My understanding is that if we somehow detect overflow/underflow and apply clipping, it would be ok?

Yes, but the only way to detect it and apply clipping is to clip in float first (at least to the range representable by int) before conversion which is basically what is done here. See also my [recent comment](https://github.com/pytorch/pytorch/pull/113861#issuecomment-1872006438) comparing this to the supposedly future quantization implementation which also does the clipping in float (i.e. before comparison) and [my comparison](https://github.com/pytorch/pytorch/pull/113861#issuecomment-1870194809) with existing quantization implementations used in PyTorch right now.

> My one concern is what you said about rounding that happens on zero point, although the code snippet I have does not quite address that.

I guess you are referring to `float(zero_point)` being potentially different than `zero_point`? This is probably rare and at least consistent to what happens currently in  `zero_point + (self * inv_scale)`

> Also your argument on bounded_qval representation in int64_t makes sense assuming qmin/qmax are exactly representable in int64_t as well.

Note that qmin/qmax have type `int64_t` so the latter is trivially true. See also [my explanation](https://github.com/pytorch/pytorch/pull/113861#issuecomment-1867453175) about the current multiple conversions between float and int looking suspicious and bad for performance and how the changed code is equivalent to the current code (at least where even currently fmin/fmax is used) in "regular" cases.

Reply:
> Lets not mix concerns of perf vs. correctness btw

Sure. My point was that it is a) suspicious in terms of clarity and correctness (i.e. it has the same results as my fixed, simpler version but may overflow) and b) has **additionally** likely worse performance, so emphasis on the "and". This means that potential performance reasons are **not** an argument to keep that code that way which sometimes does come up especially as we are talking about a fix for a corner case here.